File: src/software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play_test.cpp
#include "software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play.h"

#include <gtest/gtest.h>

#include "software/simulated_tests/simulated_er_force_sim_play_test_fixture.h"
#include "software/simulated_tests/terminating_validation_functions/friendly_scored_validation.h"
#include "software/simulated_tests/validation/validation_function.h"
#include "software/test_util/test_util.h"
#include "software/time/duration.h"
#include "software/world/world.h"

class ShootOrPassPlayTest : public SimulatedErForceSimPlayTestFixture
{
   protected:
    TbotsProto::FieldType field_type = TbotsProto::FieldType::DIV_B;
    Field field                      = Field::createField(field_type);
};

// TODO (#2577): re-enable once fixed
TEST_F(ShootOrPassPlayTest, DISABLED_test_shoot_or_pass_play)
{
    BallState ball_state(Point(-4.4, 2.9), Vector(0, 0));
    auto friendly_robots = TestUtil::createStationaryRobotStatesWithId({
        field.friendlyGoalCenter(),
        Point(-4.5, 3.0),
        Point(-2, 1.5),
        Point(-2, 0.5),
        Point(-2, -1.7),
        Point(-2, -1.5),
    });
    setFriendlyGoalie(0);
    auto enemy_robots = TestUtil::createStationaryRobotStatesWithId(
        {Point(1, 0), Point(1, 2.5), Point(1, -2.5), field.enemyGoalCenter(),
         field.enemyDefenseArea().negXNegYCorner(),
         field.enemyDefenseArea().negXPosYCorner()});
    setEnemyGoalie(0);
    setAIPlay(TYPENAME(ShootOrPassPlay));
    setRefereeCommand(RefereeCommand::FORCE_START, RefereeCommand::STOP);

    std::vector<ValidationFunction> terminating_validation_functions = {friendlyScored};

    std::vector<ValidationFunction> non_terminating_validation_functions = {};

    runTest(field_type, ball_state, friendly_robots, enemy_robots,
            terminating_validation_functions, non_terminating_validation_functions,
            Duration::fromSeconds(15));
}

File: src/software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play_test.py
import pytest

import software.python_bindings as tbots_cpp
from proto.import_all_protos import *
from proto.play_pb2 import PlayName
from proto.ssl_gc_common_pb2 import Team
from proto.message_translation.tbots_protobuf import create_world_state
from software.simulated_tests.validation.friendly_team_scored import (
    FriendlyTeamEventuallyScored,
)
from software.simulated_tests.simulated_test_fixture import pytest_main


@pytest.mark.skip(
    "Skipping test. TODO (#3233): attacker robot sometimes doesn't kick the ball towards the receiver"
)
def test_shoot_or_pass_play(simulated_test_runner):
    field = tbots_cpp.Field.createSSLDivisionBField()

    def setup(*args):
        simulated_test_runner.set_world_state(
            create_world_state(
                blue_robot_locations=[
                    field.friendlyGoalCenter(),
                    tbots_cpp.Point(-4.5, 3.0),
                    tbots_cpp.Point(-2, 1.5),
                    tbots_cpp.Point(-2, 0.5),
                    tbots_cpp.Point(-2, -1.7),
                    tbots_cpp.Point(-2, -1.5),
                ],
                yellow_robot_locations=[
                    tbots_cpp.Point(1, 0),
                    tbots_cpp.Point(1, 2.5),
                    tbots_cpp.Point(1, -2.5),
                    field.enemyGoalCenter(),
                    field.enemyDefenseArea().negXNegYCorner(),
                    field.enemyDefenseArea().negXPosYCorner(),
                ],
                ball_location=tbots_cpp.Point(-4.4, 2.9),
                ball_velocity=tbots_cpp.Vector(0, 0),
            ),
        )

        simulated_test_runner.send_gamecontroller_command(
            gc_command=Command.Type.STOP, team=Team.UNKNOWN
        )
        simulated_test_runner.send_gamecontroller_command(
            gc_command=Command.Type.FORCE_START, team=Team.BLUE
        )

        simulated_test_runner.set_plays(
            blue_play=PlayName.ShootOrPassPlay, yellow_play=PlayName.HaltPlay
        )

    # Eventually Validation
    eventually_validations = [[FriendlyTeamEventuallyScored()]]

    simulated_test_runner.run_test(
        setup=setup,
        inv_eventually_validation_sequence_set=eventually_validations,
        ag_eventually_validation_sequence_set=eventually_validations,
        test_timeout_s=15,
        run_till_end=False,
    )


if __name__ == "__main__":
    pytest_main(__file__)
